File: 브루트포스/프로그래머스lv1-모의고사.js
/*** 2021년 3월 25일의 풀이 */
function solution(answers) {
  const Children = {
    A: [1, 2, 3, 4, 5],
    B: [2, 1, 2, 3, 2, 4, 2, 5],
    C: [3, 3, 1, 1, 2, 2, 4, 4, 5, 5],
  }

  const returnScore = (arr) => {
    var i = 0
    var j = 0
    var score = 0
    while (i < answers.length) {
      if (answers[i] == arr[j]) {
        score++
      }
      i++
      j++
      if (j === arr.length) {
        j = 0
      }
    }
    return score
  }

  let score = []

  score.push(returnScore(Children.A))
  score.push(returnScore(Children.B))
  score.push(returnScore(Children.C))

  let result = []

  for (var i = 0; i < score.length; i++) {
    if (Math.max(...score) == score[i]) {
      result.push(i + 1)
    }
  }
  return result
}
/*
통과 (0.16ms, 29.9MB)
테스트 2 〉	통과 (0.23ms, 30.2MB)
테스트 3 〉	통과 (0.15ms, 30.1MB)
테스트 4 〉	통과 (0.13ms, 30.1MB)
테스트 5 〉	통과 (0.17ms, 29.9MB)
테스트 6 〉	통과 (0.17ms, 30.1MB)
테스트 7 〉	통과 (16.92ms, 32.5MB)
테스트 8 〉	통과 (0.61ms, 29.9MB)
테스트 9 〉	통과 (4.81ms, 32.8MB)
테스트 10 〉	통과 (4.70ms, 33MB)
테스트 11 〉	통과 (4.73ms, 32.3MB)
테스트 12 〉	통과 (4.90ms, 32.9MB)
테스트 13 〉	통과 (0.35ms, 29.9MB)
테스트 14 〉	통과 (4.77ms, 32.9MB)
*/

// 2021년 10월 29일, 개선된 풀이(반복문을 1번만에 끝내면서 2배 빨리진 성능)
function solution(answers) {
  const a = [1, 2, 3, 4, 5]
  const b = [2, 1, 2, 3, 2, 4, 2, 5]
  const c = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
  let score = [0, 0, 0]
  let aIdx = 0
  let bIdx = 0
  let cIdx = 0
  answers.forEach((v, i) => {
    if (v === a[aIdx++]) score[0]++
    if (v === b[bIdx++]) score[1]++
    if (v === c[cIdx++]) score[2]++

    if (aIdx === 5) aIdx = 0
    if (bIdx === 8) bIdx = 0
    if (cIdx === 10) cIdx = 0
  })

  const max = Math.max(...score)
  const result = []
  for (let i = 0; i < 3; i++) {
    if (score[i] === max) result.push(i + 1)
  }

  return result
}
/*
테스트 1 〉	통과 (0.12ms, 30.2MB)
테스트 2 〉	통과 (0.13ms, 30.3MB)
테스트 3 〉	통과 (0.10ms, 30.5MB)
테스트 4 〉	통과 (0.12ms, 30.4MB)
테스트 5 〉	통과 (0.11ms, 30.3MB)
테스트 6 〉	통과 (0.15ms, 30.3MB)
테스트 7 〉	통과 (1.84ms, 33.3MB)
테스트 8 〉	통과 (0.42ms, 30.3MB)
테스트 9 〉	통과 (2.53ms, 33.1MB)
테스트 10 〉	통과 (1.61ms, 32.9MB)
테스트 11 〉	통과 (2.55ms, 33.2MB)
테스트 12 〉	통과 (2.47ms, 32.8MB)
테스트 13 〉	통과 (0.23ms, 30.4MB)
테스트 14 〉	통과 (2.67ms, 33MB)
*/

// 여기서 더 smart 한 방법
function solution(answers) {
  const a = [1, 2, 3, 4, 5]
  const b = [2, 1, 2, 3, 2, 4, 2, 5]
  const c = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
  let score = [0, 0, 0]
  answers.forEach((v, i) => {
    if (v === a[i % 5]) score[0]++
    if (v === b[i % 8]) score[1]++
    if (v === c[i % 10]) score[2]++
  })

  const max = Math.max(...score)
  const result = []
  for (let i = 0; i < 3; i++) {
    if (score[i] === max) result.push(i + 1)
  }

  return result
}
/*
정확성  테스트
테스트 1 〉	통과 (0.08ms, 30.3MB)
테스트 2 〉	통과 (0.08ms, 30MB)
테스트 3 〉	통과 (0.08ms, 30.1MB)
테스트 4 〉	통과 (0.09ms, 30.2MB)
테스트 5 〉	통과 (0.10ms, 30.4MB)
테스트 6 〉	통과 (0.13ms, 30.4MB)
테스트 7 〉	통과 (0.76ms, 30.1MB)
테스트 8 〉	통과 (0.35ms, 30.3MB)
테스트 9 〉	통과 (1.86ms, 33.2MB)
테스트 10 〉	통과 (0.63ms, 30.4MB)
테스트 11 〉	통과 (3.33ms, 33MB)
테스트 12 〉	통과 (1.81ms, 32.9MB)
테스트 13 〉	통과 (0.20ms, 30.3MB)
테스트 14 〉	통과 (1.98ms, 33.2MB)
*/

// 2022년 7월 풀이
// 위 풀이들처럼 하나하나 인덱스로 접근하면 당연히 시간복잡도를 줄일 수 있다.
// 문제와는 관련이 없지만 학생들이 많아질 경우 저렇게 코드를 짤 수는 없음. 따라서 어쩔 수없이 2중 루프를 선택해야 할 것이다.
function solution(answers) {
  const N = 3
  const table = [
    [1, 2, 3, 4, 5],
    [2, 1, 2, 3, 2, 4, 2, 5],
    [3, 3, 1, 1, 2, 2, 4, 4, 5, 5],
  ]

  const result = Array.from({ length: N + 1 }, (v, i) => 0)

  answers.forEach((answer, i) => {
    table.forEach((x, j) => {
      if (x[i % x.length] === answer) result[j + 1]++
    })
  })

  const mostValue = Math.max(...result)
  const iter = Array.from({ length: N }, (v, i) => i + 1)

  // 이미 정렬되어있으므로 다시 정렬시킬 필요는 없음.
  return iter.reduce((acc, cur) => {
    if (mostValue === result[cur]) {
      acc.push(cur)
    }
    return acc
  }, [])
}

/*
정확성 테스트
테스트 1 〉 통과 (0.22ms, 33.4MB)
테스트 2 〉 통과 (0.21ms, 33.4MB)
테스트 3 〉 통과 (0.13ms, 33.5MB)
테스트 4 〉 통과 (0.13ms, 33.4MB)
테스트 5 〉 통과 (0.34ms, 33.5MB)
테스트 6 〉 통과 (0.26ms, 33.5MB)
테스트 7 〉 통과 (1.96ms, 36.3MB)
테스트 8 〉 통과 (0.60ms, 33.7MB)
테스트 9 〉 통과 (4.19ms, 36.7MB)
테스트 10 〉 통과 (2.03ms, 36.3MB)
테스트 11 〉 통과 (2.83ms, 36.8MB)
테스트 12 〉 통과 (3.19ms, 37.2MB)
테스트 13 〉 통과 (0.34ms, 33.5MB)
테스트 14 〉 통과 (3.04ms, 36.9MB)

*/

File: 브루트포스/프로그래머스lv1-모의고사.py
def solution(answers):
    a = [1, 2, 3, 4, 5]
    b = [2, 1, 2, 3, 2, 4, 2, 5]
    c = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    score = [0, 0, 0]
    a_i = 0
    b_i = 0
    c_i = 0
    for i, answer in enumerate(answers):
        if answer == a[a_i]:
            score[0] += 1
        if answer == b[b_i]:
            score[1] += 1
        if answer == c[c_i]:
            score[2] += 1
        
        a_i += 1
        b_i += 1
        c_i += 1
            
        if a_i == 5:
            a_i = 0
        if b_i == 8:
            b_i = 0
        if c_i == 10:
            c_i = 0
        
    _max = max(score)
    result = []
    for i in range(3):
        if score[i] == _max:
            result.append(i + 1)
    
    return result

"""
테스트 1 〉	통과 (0.01ms, 10.3MB)
테스트 2 〉	통과 (0.01ms, 10.4MB)
테스트 3 〉	통과 (0.01ms, 10.3MB)
테스트 4 〉	통과 (0.01ms, 10.2MB)
테스트 5 〉	통과 (0.02ms, 10.4MB)
테스트 6 〉	통과 (0.02ms, 10.3MB)
테스트 7 〉	통과 (1.61ms, 10.3MB)
테스트 8 〉	통과 (0.45ms, 10.2MB)
테스트 9 〉	통과 (2.32ms, 10.3MB)
테스트 10 〉	통과 (1.10ms, 10.3MB)
테스트 11 〉	통과 (2.45ms, 10.4MB)
테스트 12 〉	통과 (2.55ms, 10.3MB)
테스트 13 〉	통과 (0.14ms, 10.3MB)
테스트 14 〉	통과 (2.49ms, 10.3MB)
"""

# 조금 더 smart한 방법 1 -> 딕셔너리
def solution(answers):
    a = [1, 2, 3, 4, 5]
    b = [2, 1, 2, 3, 2, 4, 2, 5]
    c = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]

    counts = {
        1: 0,
        2: 0,
        3: 0
    }
    for i in range(len(answers)):
        if answers[i] == a[i % len(a)]:
            counts[1] += 1
        if answers[i] == b[i % len(b)]:
            counts[2] += 1
        if answers[i] == c[i % len(c)]:
            counts[3] += 1
    
    _max = max(counts.values())
    result = [x[0] for x in counts.items() if x[1] == _max]
    # sorted(result)를 안해도 무방
    return result

# 조금 더 smart한 방법 1 -> 배열
def solution(answers):
    a = [1, 2, 3, 4, 5]
    b = [2, 1, 2, 3, 2, 4, 2, 5]
    c = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]

    counts = [0] * 3
    for i in range(len(answers)):
        if answers[i] == a[i % len(a)]:
            counts[0] += 1
        if answers[i] == b[i % len(b)]:
            counts[1] += 1
        if answers[i] == c[i % len(c)]:
            counts[2] += 1

    _max = max(counts)
    result = []
    for i in range(3):
        if counts[i] == _max:
            result.append(i + 1)
    # sorted(result) 를 생략해도 무방
    return result

"""
두 결과 모두 비슷하다.

정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.3MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.01ms, 10.3MB)
테스트 4 〉	통과 (0.01ms, 10.3MB)
테스트 5 〉	통과 (0.02ms, 10.3MB)
테스트 6 〉	통과 (0.02ms, 10.2MB)
테스트 7 〉	통과 (1.09ms, 10.4MB)
테스트 8 〉	통과 (0.35ms, 10.2MB)
테스트 9 〉	통과 (2.00ms, 10.3MB)
테스트 10 〉	통과 (0.90ms, 10.3MB)
테스트 11 〉	통과 (2.17ms, 10.3MB)
테스트 12 〉	통과 (1.77ms, 10.3MB)
테스트 13 〉	통과 (0.12ms, 10.3MB)
테스트 14 〉	통과 (2.38ms, 10.3MB)
"""

# 2022년 7월 풀이
# 위 풀이들처럼 하나하나 인덱스로 접근하면 당연히 시간복잡도를 줄일 수 있다.
# 문제와는 관련이 없지만 학생들이 많아질 경우 저렇게 코드를 짤 수는 없음. 따라서 어쩔 수없이 2중 루프를 선택해야 할 것이다.
def solution(answers):
    N = 3
    table = [
        [],
        [1, 2, 3, 4, 5],
        [2, 1, 2, 3, 2, 4, 2, 5],
        [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]
    ]
    result = [0] * (N + 1)

    for i, answer in enumerate(answers):
        for j in range(1, N + 1):
            x = table[j]
            if x[i % len(x)] == answer:
                result[j] += 1
    
    most_value = max(result)

    return sorted([i for i in range(1, N + 1) if result[i] == most_value])
"""
정확성 테스트
테스트 1 〉 통과 (0.01ms, 10.2MB)
테스트 2 〉 통과 (0.01ms, 10.3MB)
테스트 3 〉 통과 (0.01ms, 10.1MB)
테스트 4 〉 통과 (0.01ms, 10.3MB)
테스트 5 〉 통과 (0.04ms, 10.1MB)
테스트 6 〉 통과 (0.05ms, 10.2MB)
테스트 7 〉 통과 (2.98ms, 10.1MB)
테스트 8 〉 통과 (0.94ms, 10.2MB)
테스트 9 〉 통과 (6.12ms, 10MB)
테스트 10 〉 통과 (2.33ms, 10.4MB)
테스트 11 〉 통과 (9.23ms, 10.4MB)
테스트 12 〉 통과 (5.34ms, 10.4MB)
테스트 13 〉 통과 (0.29ms, 10.1MB)
테스트 14 〉 통과 (5.63ms, 10.4MB)
"""